modules/security-best-practices/pages/web-app-development-security-best-practices.adoc

```diff
@@ -0,0 +1,126 @@
+= Web App Development Security Best Practices
+
+== Authentication
+
+=== Use a well-audited authentication service and client side IC libraries
+
+==== Security Concern
+
+Implementing user authentication and canister calls yourself in your web app is error prone and risky. For example, if canister calls are implemented from scratch, there may be bugs around signature creation or verification.
+
+==== Recommendation
+
+- Consider using e.g. link:https://github.com/dfinity/internet-identity[Internet Identity] for authentication, use link:https://github.com/dfinity/agent-js[agent-js] for making canister calls, and the link:https://github.com/dfinity/agent-js/tree/main/packages/auth-client[auth-client] for interacting with internet identity from your dApp.
+- It is of course also an option to consider alternative authentication frameworks on the IC for authentication.
+
+=== Set an appropriate session timeout
+
+==== Security Concern
+
+Currently, Internet Identity issues delegations with an expiry time. This expiry time can be set in the link:https://github.com/dfinity/agent-js/tree/main/packages/auth-client[auth-client]. After a delegation expires, the user has to re-authenticate. Setting a good value is a trade-off between security and usability.
+
+==== Recommendation
+
+See the link:https://cheatsheetseries.owasp.org/cheatsheets/Session_Management_Cheat_Sheet.html#session-expiration[OWASP Recommendations]. A timeout of 30min should be set for security sensitive applications.
+
+Note: Idle timeouts might be introduced soon, see link:https://forum.dfinity.org/t/authclient-update-idle-timeouts/10464[this].
+
+=== Don’t use fetchRootKey in agent-js in production
+
+==== Security Concern
+
+`agent.fetchRootKey()` can be used in link:https://github.com/dfinity/agent-js[agent-js] to fetch the root subnet threshold public key from a status call in test environments. This key is used to verify threshold signatures on certified data received through canister update calls. Using this method in a production web app gives an attacker the option to supply their own public key, invalidating all authenticity guarantees of update responses.
+
+==== Recommendation
+
+Never use `agent.fetchRootKey()`only in test builds, but not in production builds. Not calling this method will result in the hard coded root subnet public key of mainnet being used for signature verification, which is the desired behavior in production.
```

Comment:
```suggestion
Never use `agent.fetchRootKey()` in production builds, only in test builds. Not calling this method will result in the hard coded root subnet public key of mainnet being used for signature verification, which is the desired behavior in production.
```

Reply:
Correct, thanks for spotting this @sesi200 ! Fixed it.